src/iceberg/table_scan.h

```diff
@@ -0,0 +1,204 @@
+};
+
+/// \brief Scan context holding snapshot and scan-specific metadata.
+struct TableScanContext {
+  std::shared_ptr<TableMetadata> table_metadata;  ///< Table metadata.
+  std::shared_ptr<Snapshot> snapshot;             ///< Snapshot to scan.
+  std::shared_ptr<Schema> projected_schema;       ///< Projected schema.
+  std::shared_ptr<Expression> filter;             ///< Filter expression to apply.
+  bool case_sensitive = false;                    ///< Whether the scan is case-sensitive.
+  std::unordered_map<std::string, std::string>
+      options;                   ///< Additional options for the scan.
+  std::optional<int64_t> limit;  ///< Optional limit on the number of rows to scan.
+};
+
+/// \brief Builder class for creating TableScan instances.
+class ICEBERG_EXPORT TableScanBuilder {
+ public:
+  /// \brief Constructs a TableScanBuilder for the given table.
+  /// \param table The table to scan.
+  /// \param table_metadata The metadata of the table to scan.
+  explicit TableScanBuilder(const Table& table,
+                            std::shared_ptr<TableMetadata> table_metadata);
+
+  /// \brief Sets the snapshot ID to scan.
+  /// \param snapshot_id The ID of the snapshot.
+  /// \return Reference to the builder.
+  TableScanBuilder& WithSnapshotId(int64_t snapshot_id);
+
+  /// \brief Selects columns to include in the scan.
+  /// \param column_names A list of column names. If empty, all columns will be selected.
+  /// \return Reference to the builder.
+  TableScanBuilder& WithColumnNames(std::vector<std::string> column_names);
+
+  /// \brief Sets the schema to use for the scan.
+  /// \param schema The schema to use.
+  /// \return Reference to the builder.
+  TableScanBuilder& WithSchema(std::shared_ptr<Schema> schema);
+
+  /// \brief Applies a filter expression to the scan.
+  /// \param filter Filter expression to use.
+  /// \return Reference to the builder.
+  TableScanBuilder& WithFilter(std::shared_ptr<Expression> filter);
+
+  /// \brief Sets whether the scan should be case-sensitive.
+  /// \param case_sensitive Whether the scan is case-sensitive.
+  /// /return Reference to the builder.
+  TableScanBuilder& WithCaseSensitive(bool case_sensitive);
+
+  /// \brief Sets an option for the scan.
+  /// \param property The name of the option.
+  /// \param value The value of the option.
+  /// \return Reference to the builder.
+  TableScanBuilder& WithOption(std::string property, std::string value);
+
+  /// \brief Sets an optional limit on the number of rows to scan.
+  /// \param limit Optional limit on the number of rows.
+  /// \return Reference to the builder.
+  TableScanBuilder& WithLimit(std::optional<int64_t> limit);
+
+  /// \brief Builds and returns a TableScan instance.
+  /// \return A Result containing the TableScan or an error.
+  Result<std::unique_ptr<TableScan>> Build();
+
+ private:
+  const Table& table_;  ///< Reference to the table to scan.
+  std::vector<std::string> column_names_;
```

Comment:
Why there is no `std::shared_ptr<Schema>` for `WithSchema`? Same question for filter.